### PR TITLE
Idsvr uriprovider

### PR DESCRIPTION
**Pull request by AroglDarthu**
**What this PR does / why we need it:**
Add overload to IdSvrHealthCheck which enables the uri to be resolved from the IServiceProvider.

**Does this PR introduce a user-facing change?:**
No

**Please make sure you've completed the relevant tasks for this PR, out of the following list:**

[ x] Code compiles correctly
[ x] Created/updated tests
[ x] Unit tests passing
[ x] End-to-end tests passing
[ -] Extended the documentation
[ -] Provided sample for the feature

### CODE REVIEW

**Comment by unaizorrilla** [comment]
Hi @AroglDarthu

Can you explain with more details this scenario? Typically the authority uri is con your configuration and you can use IConfiguration when registered the healtcheck. What kind of scenaris is enabled using the Func<IServiceProvider,Uri>??

**Comment by AroglDarthu** [comment]
We have dedicated objects that read from configuration. These get
registered via "services.Configure<DedicatedPoco>". So we can just inject
these wherever needed. We prefer to also use these in the healthcheck
instead of giving the healthcheck the knowledge about how to obtain the
values from IConfiguration.

Op vr 25 sep. 2020 13:11 schreef Unai Zorrilla <notifications@github.com>:

**Comment by sungam3r** [comment]
@AroglDarthu Please rebase onto `master` if you are going to continue work on this PR. Then I can review/merge.

**Comment by AroglDarthu** [comment]
> @AroglDarthu Please rebase onto `master` if you are going to continue work on this PR. Then I can review/merge.

That would be nice. I'll look into it as soon as I can.

**Comment by sungam3r** [comment]
ping @AroglDarthu

**Comment by AroglDarthu** [comment]
@sungam3r sorry for the long wait!

File: src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs [hunk 1]
```diff
@@ -42,5 +42,45 @@ public static class IdSvrHealthCheckBuilderExtensions
...
+
+            return builder.Add(
+                new HealthCheckRegistration(
+                    registrationName,
+                    sp => new IdSvrHealthCheck(() => CreateIdentityServerHttpClient(sp, registrationName, uriProvider)),
+                    failureStatus,
+                    tags,
+                    timeout));
+        }
+        private static HttpClient CreateIdentityServerHttpClient(IServiceProvider sp, string registrationName, Func<IServiceProvider, Uri> uriProvider)
```
Comment (sungam3r) [inline src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs:76]:
```suggestion
        private static HttpClient CreateIdentityServerHttpClient(IServiceProvider sp, string registrationName, Func<IServiceProvider, Uri> uriProvider)
```
```suggestion


        private static HttpClient CreateIdentityServerHttpClient(IServiceProvider sp, string registrationName, Func<IServiceProvider, Uri> uriProvider)
```

File: src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs [hunk 2]
```diff
@@ -42,5 +42,47 @@ public static class IdSvrHealthCheckBuilderExtensions
...
+        /// <returns>The specified <paramref name="builder"/>.</returns>
+        public static IHealthChecksBuilder AddIdentityServer(
+            this IHealthChecksBuilder builder,
+            Func<IServiceProvider, Uri> uriProvider,
+            string name = null,
+            HealthStatus? failureStatus = null,
+            IEnumerable<string> tags = null,
+            TimeSpan? timeout = null)
+        {
+            builder.Services.AddHttpClient();
```
Comment (sungam3r) [inline src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs:65]: `AddHttpClient` has overload with service provider. I would you it to configure http client, get rid of `CreateIdentityServerHttpClient` method and keep `builder.Add` call the same as above.

File: src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs [hunk 3]
```diff
@@ -42,5 +42,47 @@ public static class IdSvrHealthCheckBuilderExtensions
...
+
+            return builder.Add(
+                new HealthCheckRegistration(
+                    registrationName,
+                    sp => new IdSvrHealthCheck(() => CreateIdentityServerHttpClient(sp, registrationName, uriProvider)),
+                    failureStatus,
+                    tags,
+                    timeout));
+        }
+
```
Comment (sungam3r) [inline src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs:76]:
```suggestion
```

File: src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -42,5 +42,49 @@ public static class IdSvrHealthCheckBuilderExtensions
...
+        }
+        private static HttpClient CreateIdentityServerHttpClient(IServiceProvider sp, string registrationName, Func<IServiceProvider, Uri> uriProvider)
+        {
+            var authorityUri = uriProvider(sp);
+            var httpClientFactory = sp.GetRequiredService<IHttpClientFactory>();
+            var client = httpClientFactory.CreateClient(registrationName);
+            client.BaseAddress = authorityUri;
+
+            return client;
+        }
```
Comment (sungam3r) [inline src/HealthChecks.OpenIdConnectServer/DependencyInjection/IdSvrHealthCheckBuilderExtensions.cs:88]:
```suggestion
```

**Comment by sungam3r** [comment]
The last step - update API approval file.

**Review by sungam3r — APPROVED** [review]

File: test/HealthChecks.OpenIdConnectServer.Tests/HealthChecks.OpenIdConnectServer.approved.txt [hunk 5]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.Extensions.DependencyInjection
 {
     public static class IdSvrHealthCheckBuilderExtensions
     {
+        public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddIdentityServer(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, System.Func<System.IServiceProvider, System.Uri> uriProvider, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
         public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddIdentityServer(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, System.Uri idSvrUri, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
     }
-}
\ No newline at end of file
+}
```
Comment (sungam3r) [inline test/HealthChecks.OpenIdConnectServer.Tests/HealthChecks.OpenIdConnectServer.approved.txt:16]:
```suggestion
}

```

**Comment by sungam3r** [comment]
@AroglDarthu Problem with line endings.

File: test/HealthChecks.OpenIdConnectServer.Tests/HealthChecks.OpenIdConnectServer.approved.txt (hunk 5 above)
Comment (AroglDarthu) [reply test/HealthChecks.OpenIdConnectServer.Tests/HealthChecks.OpenIdConnectServer.approved.txt]: My editor kept insisting on adding the empty line. `cp` to the rescue.

**Pull request merged**